libevmasm/Assembly.cpp

```diff
@@ -74,6 +75,130 @@ unsigned Assembly::codeSize(unsigned subTagSize) const
 	}
 }
 
+void Assembly::addAssemblyItemsFromJSON(Json::Value const& _code)
+{
+	solAssert(m_items.empty());
+	solAssert(_code.isArray());
+	for (auto const& jsonItem: _code)
+		m_items.emplace_back(createAssemblyItemFromJSON(jsonItem));
+
+	for (auto current = m_items.begin(); current != m_items.end(); ++current)
+	{
+		// During the assembly json export a `JUMPDEST` is always generated after a `tag`.
+		// So we just ignore exactly these `JUMPDEST`'s.
+		auto const nextItem = next(current);
+		if (nextItem != m_items.end() &&
+			current->type() == AssemblyItemType::Tag &&
+			nextItem->type() == AssemblyItemType::Operation &&
+			nextItem->instruction() == Instruction::JUMPDEST
+		)
+			m_items.erase(nextItem);
+	}
+
+	for (auto const& item: m_items)
+		if (item.type() == AssemblyItemType::Operation && item.instruction() == Instruction::JUMPDEST)
+			throw langutil::Error(
+				1285_error,
+				langutil::Error::Type::FatalError,
+				"JUMPDEST instruction found that was not followed by tag."
+			);
+}
+
+AssemblyItem Assembly::createAssemblyItemFromJSON(Json::Value const& _json)
+{
+	solAssert(isOfType<string>(_json["name"]));
+	solAssert(isOfType<int>(_json["begin"]));
+	solAssert(isOfType<int>(_json["end"]));
+	solAssert(isOfType<int>(_json["source"]));
+	solAssert(isOfTypeIfExists<string>(_json, "value"));
+	solAssert(isOfTypeIfExists<int>(_json, "modifierDepth"));
+	solAssert(isOfTypeIfExists<string>(_json, "jumpType"));
```

Comment:
What do you mean with older fields? With unknown fields I think we should just ignore them, maybe a tool wants to add additional data in some other fields.

Reply:
Sorry, I meant "other". Not sure why it ended up as "older" :)

But back to the point, I think validation should be strict unless there's a good reason to expect there may be some unknown fields there. if extra fields are ignored, it's not easy to notice that you're using the format incorrectly. I actually remember a similar case from not long ago - people from dapptools had a really hard time figuring out the right format for library references for the linker. The compiler was unhelpful because it just took the invalid format ran with it, but was not using those references. It would be more helpful to get an error saying what's wrong with the input.

A good reason to have a more relaxed validation would be if we would have to worry about supporting different compiler versions. But I don't think we do. The assembly output is highly version specific (different version will very likely give you slightly different instructions due to changes in the optimizer) so you don't really want to store output from one version and reuse it later. You want to regenerate it from the source again.

Another one might be the tool having its own format of which we only use a subset. I think it's not the case here. The tool can also always convert its richer format to one that only contains info relevant to the compiler.

Reply:
Ok, yeah makes sense. So let's be more strict here. If someone needs to have it less strict, we can also just change that later. I agree that it would be easier to use, if a stricter check is in place. Especially if the user does not know the format.

Reply:
Added a check for valid members `{"name", "begin", "end", "source", "value", "modifierDepth", "jumpType"}` everything else will create an `AssemblyImportException`.

Reply:
But I also changed that some of the attributes can be now optional. This is quite handy e.g. while writing the command-line tests for this PR.